### PR TITLE
Add the plist resource from the macos cookbook

### PR DESCRIPTION
**Pull request by tas50**
This is a very useful base resource for mac systems that allows us to
write additional resources for managing macs.

Signed-off-by: Tim Smith <tsmith@chef.io>

### CODE REVIEW

File: lib/chef/resource/plist.rb [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+        file_group_cmd = shell_out("/usr/bin/stat", "-f", "%Sg", desired.path)
+        group file_group_cmd.stdout.chomp
+      end
+
+      action :set do
+        converge_if_changed :path do
+          converge_by "create new plist: '#{new_resource.path}'" do
+            file new_resource.path do
+              empty_plist = {}.to_plist
+              content empty_plist
```
Comment (phiggins) [inline lib/chef/resource/plist.rb:62]:
```suggestion 
              content {}.to_plist
```

File: lib/chef/resource/plist.rb [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        cmd.exitstatus == 0
+      end
+
+      def plistbuddy_command(subcommand, entry, path, value = nil)
+        sep = " "
+        arg = case subcommand.to_s
+              when "add"
+                type_to_commandline_string(value)
+              when "set"
+                if value.class == Hash
```
Comment (phiggins) [inline lib/chef/resource/plist.rb:172]:
```suggestion
                if value.is_a?(Hash)
```

File: lib/chef/resource/plist.rb [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+        "/usr/bin/plutil"
+      end
+
+      def defaults_executable
+        "/usr/bin/defaults"
+      end
+
+      def plistbuddy_executable
+        "/usr/libexec/PlistBuddy"
+      end
```
Comment (phiggins) [inline lib/chef/resource/plist.rb:216]: These should probably all be constants.

File: spec/unit/resource/plist_spec.rb [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+      it "returns the required integer entry type as a string" do
+        expect(resource.type_to_commandline_string(1)).to eq "integer"
+      end
+
+      it "returns the required float entry type as a string" do
+        expect(resource.type_to_commandline_string(1.0)).to eq "float"
+      end
+    end
+  end
+end
```
Comment (phiggins) [inline spec/unit/resource/plist_spec.rb:154]: There are a lot of superfluous `context`s in here that don't really add much IMO.

File: lib/chef/resource/plist.rb (hunk 3 above)
Comment (tas50) [reply lib/chef/resource/plist.rb]: Done

File: spec/unit/resource/plist_spec.rb (hunk 4 above)
Comment (tas50) [reply spec/unit/resource/plist_spec.rb:133]: Nuked most of them

File: spec/unit/resource/plist_spec.rb [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+      expect(resource.plistbuddy_command(:add, "Foo Bar Baz", "path/to/file.plist", true)).to eq "/usr/libexec/PlistBuddy -c 'Add :\"Foo Bar Baz\" bool' \"path/to/file.plist\""
+    end
+
+    it "returns the value properly formatted with double quotes when the plist itself contains spaces" do
+      expect(resource.plistbuddy_command(:print, "Foo Bar Baz", "Library/Preferences/com.parallels.Parallels Desktop.plist")).to eq "/usr/libexec/PlistBuddy -c 'Print :\"Foo Bar Baz\"' \"Library/Preferences/com.parallels.Parallels Desktop.plist\""
+    end
+  end
+
+  describe "#convert_to_data_type_from_string" do
+
```
Comment (phiggins) [inline spec/unit/resource/plist_spec.rb:74]:
```suggestion
```

File: spec/unit/resource/plist_spec.rb [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+      end
+
+      it "the print command is formatted properly" do
+        expect(resource.plistbuddy_command(:print, "QuxEntry", "path/to/file.plist")).to eq "/usr/libexec/PlistBuddy -c 'Print :\"QuxEntry\"' \"path/to/file.plist\""
+      end
+
+      it "the command to set a dictionary data type is formatted properly" do
+        expect(resource.plistbuddy_command(:set, "AppleFirstWeekday", "path/to/file.plist", gregorian: 4)).to eq "/usr/libexec/PlistBuddy -c 'Set :\"AppleFirstWeekday\":gregorian 4' \"path/to/file.plist\""
+      end
+    end
```
Comment (phiggins) [inline spec/unit/resource/plist_spec.rb:58]:
```suggestion
    it "the bool arguments contain the data type" do
      expect(resource.plistbuddy_command(:add, "FooEntry", "path/to/file.plist", true)).to eq "/usr/libexec/PlistBuddy -c 'Add :\"FooEntry\" bool' \"path/to/file.plist\""
    end

    it "the add command only adds the data type" do
      expect(resource.plistbuddy_command(:add, "QuuxEntry", "path/to/file.plist", 50)).to eq "/usr/libexec/PlistBuddy -c 'Add :\"QuuxEntry\" integer' \"path/to/file.plist\""
    end

    it "the delete command is formatted properly" do
      expect(resource.plistbuddy_command(:delete, "BarEntry", "path/to/file.plist")).to eq "/usr/libexec/PlistBuddy -c 'Delete :\"BarEntry\"' \"path/to/file.plist\""
    end

    it "the set command is formatted properly" do
      expect(resource.plistbuddy_command(:set, "BazEntry", "path/to/file.plist", false)).to eq "/usr/libexec/PlistBuddy -c 'Set :\"BazEntry\" false' \"path/to/file.plist\""
    end

    it "the print command is formatted properly" do
      expect(resource.plistbuddy_command(:print, "QuxEntry", "path/to/file.plist")).to eq "/usr/libexec/PlistBuddy -c 'Print :\"QuxEntry\"' \"path/to/file.plist\""
    end

    it "the command to set a dictionary data type is formatted properly" do
      expect(resource.plistbuddy_command(:set, "AppleFirstWeekday", "path/to/file.plist", gregorian: 4)).to eq "/usr/libexec/PlistBuddy -c 'Set :\"AppleFirstWeekday\":gregorian 4' \"path/to/file.plist\""
    end
```

File: lib/chef/resource/plist.rb [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+            file new_resource.path do
+              content {}.to_plist
+              owner new_resource.owner
+              group new_resource.group
+              mode new_resource.mode if property_is_set?(:mode)
+            end
+          end
+        end
+
+        plist_file_name = new_resource.path.split("/").last
```
Comment (phiggins) [inline lib/chef/resource/plist.rb:77]:
```suggestion
        plist_file_name = ::File.basename(new_resource.path)
```

**Pull request merged**